### PR TITLE
SQLite: Foreign Key Support

### PR DESCRIPTION
**Pull request by kamipo**
https://www.sqlite.org/foreignkeys.html

### CODE REVIEW

File: activerecord/test/schema/schema.rb [hunk 1]
```diff
@@ -992,16 +992,19 @@
   create_table :records, force: true do |t|
   end
 
-  if supports_foreign_keys?
-    # fk_test_has_fk should be before fk_test_has_pk
-    create_table :fk_test_has_fk, force: true do |t|
-      t.integer :fk_id, null: false
-    end
+  if supports_foreign_keys_in_create?
```
Comment (sgrif) [inline activerecord/test/schema/schema.rb]: We shouldn't be pushing caring about this onto our users. If `create_table` is called with foreign keys on SQLite, we should do a separate query automatically

Comment (kamipo) [inline activerecord/test/schema/schema.rb]: If `create_table` is called with foreign keys, generates in single query. ref #20009.

Comment (sgrif) [inline activerecord/test/schema/schema.rb]: I know. You're missing my point. It needs to not do that on SQLite. This code should be capable of running on both backends.

Comment (kamipo) [inline activerecord/test/schema/schema.rb]: Sorry I don't understand. AFAIK this code block works on mysql2, postgresql, and sqlite3 adapters.

```ruby
  if supports_foreign_keys_in_create?
    drop_table :fk_test_has_fk, if_exists: true

    create_table :fk_test_has_pk, force: true, primary_key: "pk_id" do |t|
    end

    create_table :fk_test_has_fk do |t|
      t.integer :fk_id, null: false
      t.foreign_key :fk_test_has_pk, column: "fk_id", name: "fk_name", primary_key: "pk_id"
    end
  end
```

MySQL:

```sql
DROP TABLE IF EXISTS `fk_test_has_fk`
DROP TABLE `fk_test_has_pk`
CREATE TABLE `fk_test_has_pk` (`pk_id` int AUTO_INCREMENT PRIMARY KEY) ENGINE=InnoDB
CREATE TABLE `fk_test_has_fk` (`id` int AUTO_INCREMENT PRIMARY KEY, `fk_id` int NOT NULL, CONSTRAINT `fk_name`
FOREIGN KEY (`fk_id`)
  REFERENCES `fk_test_has_pk` (`pk_id`)
) ENGINE=InnoDB
```

PostgreSQL:

```sql
DROP TABLE IF EXISTS "fk_test_has_fk"
DROP TABLE "fk_test_has_pk"
CREATE TABLE "fk_test_has_pk" ("pk_id" serial primary key)
CREATE TABLE "fk_test_has_fk" ("id" serial primary key, "fk_id" integer NOT NULL, CONSTRAINT "fk_name"
FOREIGN KEY ("fk_id")
  REFERENCES "fk_test_has_pk" ("pk_id")
)
```

SQLite3:

```sql
DROP TABLE IF EXISTS "fk_test_has_fk"
DROP TABLE "fk_test_has_pk"
CREATE TABLE "fk_test_has_pk" ("pk_id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL)
CREATE TABLE "fk_test_has_fk" ("id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, "fk_id" integer NOT NULL, CONSTRAINT "fk_name"
FOREIGN KEY ("fk_id")
  REFERENCES "fk_test_has_pk" ("pk_id")
)
```

Comment (sgrif) [inline activerecord/test/schema/schema.rb]: What I'm saying is this should work without the if statement. Pn versions of SQLite that do not support foreign keys in create, or any other adapter that doesn't support this as a single query, it should automatically perform 2 queries.

File: activerecord/test/schema/schema.rb [hunk 2]
```diff
@@ -992,16 +992,17 @@
...
-    create_table :fk_test_has_pk, force: true, primary_key: "pk_id" do |t|
-    end
+  create_table :fk_test_has_pk, force: true, primary_key: "pk_id" do |t|
+  end
 
-    add_foreign_key :fk_test_has_fk, :fk_test_has_pk, column: "fk_id", name: "fk_name", primary_key: "pk_id"
+  create_table :fk_test_has_fk do |t|
+    t.integer :fk_id, null: false
+    t.foreign_key :fk_test_has_pk, column: "fk_id", name: "fk_name", primary_key: "pk_id"
+  end
```
Comment (kamipo) [inline activerecord/test/schema/schema.rb]: @sgrif Thank you for explanation, I understand. Actually this works without `if supports_foreign_keys_in_create?`. I removed the if statement.

Comment (sgrif) [inline activerecord/test/schema/schema.rb]: Tested on an older version of SQLite?

Comment (kamipo) [inline activerecord/test/schema/schema.rb]:
```ruby
def (ActiveRecord::Base.connection).sqlite_version
  @sqlite_version ||= ActiveRecord::ConnectionAdapters::SQLite3Adapter::Version.new('3.6.16')
end

ActiveRecord::Schema.define do
  drop_table :fk_test_has_fk, if_exists: true

  create_table :fk_test_has_pk, force: true, primary_key: "pk_id" do |t|
  end

  create_table :fk_test_has_fk do |t|
    t.integer :fk_id, null: false
    t.foreign_key :fk_test_has_pk, column: "fk_id", name: "fk_name", primary_key: "pk_id"
  end
end
```

Result:

```sql
DROP TABLE IF EXISTS "fk_test_has_fk"
DROP TABLE "fk_test_has_pk"
CREATE TABLE "fk_test_has_pk" ("pk_id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL)
CREATE TABLE "fk_test_has_fk" ("id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, "fk_id" integer NOT NULL)
```

https://github.com/kamipo/rails/blob/sqlite_foreing_key_support/activerecord/lib/active_record/connection_adapters/abstract/schema_creation.rb#L52-L54
https://github.com/kamipo/rails/blob/sqlite_foreing_key_support/activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb#L910

Even Ubuntu 10.04 bundles sqlite3 3.6.22. To test an actual older version (< 3.6.19) is difficult for me...
https://launchpad.net/ubuntu/+source/sqlite3/3.6.22-1

File: activerecord/test/schema/schema.rb [hunk 3]
```diff
@@ -992,19 +992,18 @@
...
-    create_table :fk_test_has_pk, force: true, primary_key: "pk_id" do |t|
-    end
+  create_table :fk_test_has_pk, force: true, primary_key: "pk_id" do |t|
+  end
 
-    add_foreign_key :fk_test_has_fk, :fk_test_has_pk, column: "fk_id", name: "fk_name", primary_key: "pk_id"
-    add_foreign_key :lessons_students, :students
+  create_table :fk_test_has_fk do |t|
+    t.integer :fk_id, null: false
+    t.foreign_key :fk_test_has_pk, column: "fk_id", name: "fk_name", primary_key: "pk_id"
```
Comment (kamipo) [inline activerecord/test/schema/schema.rb]: > Tested on an older version of SQLite?

```ruby
def (ActiveRecord::Base.connection).sqlite_version
  @sqlite_version ||= ActiveRecord::ConnectionAdapters::SQLite3Adapter::Version.new('3.6.16')
end

ActiveRecord::Schema.define do
  drop_table :fk_test_has_fk, if_exists: true

  create_table :fk_test_has_pk, force: true, primary_key: "pk_id" do |t|
  end

  create_table :fk_test_has_fk do |t|
    t.integer :fk_id, null: false
    t.foreign_key :fk_test_has_pk, column: "fk_id", name: "fk_name", primary_key: "pk_id"
  end
end
```

Result:

```sql
DROP TABLE IF EXISTS "fk_test_has_fk"
DROP TABLE "fk_test_has_pk"
CREATE TABLE "fk_test_has_pk" ("pk_id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL)
CREATE TABLE "fk_test_has_fk" ("id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, "fk_id" integer NOT NULL)
```

https://github.com/kamipo/rails/blob/sqlite_foreing_key_support/activerecord/lib/active_record/connection_adapters/abstract/schema_creation.rb#L52-L54
https://github.com/kamipo/rails/blob/sqlite_foreing_key_support/activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb#L910

Even Ubuntu 10.04 bundles sqlite3 3.6.22. To test an actual older version (< 3.6.19) is difficult for me...
https://launchpad.net/ubuntu/+source/sqlite3/3.6.22-1

**Comment by kamipo** [comment]
Since you has merged #27701, and #24986 has assigned to you, can you also review the PR for foreign key creation? r? @pixeltrix

**Comment by pixeltrix** [comment]
@kamipo doesn't this include the changes in #24986

**Comment by pixeltrix** [comment]
@kamipo thanks! 👍

**Pull request merged**